### PR TITLE
feat(dnn) add nchw44 deconv

### PR DESCRIPTION
**Pull request by wangxiang9603**
这是这个项目的代码：https://summer-ospp.ac.cn/#/org/prodetail/222940130

### CODE REVIEW

**Comment by wangxiang9603** [comment]
@chenqy4933 please review

**Comment by chenqy4933** [comment]
@1695924908 帮忙一起review一下

**Comment by chenqy4933** [comment]
@yeasoon 帮忙一起review一下

File: dnn/src/common/convolution.cpp [hunk 1]
```diff
@@ -1156,7 +1156,7 @@ void ConvolutionBackwardData::deduce_layout(
     MEGDNN_MARK_USED_VAR(errmsg);
     megdnn_assert_contiguous(filter);
     megdnn_assert_contiguous(diff);
-    megdnn_assert(filter.ndim == 4_z || filter.ndim == 5_z, "%s", errmsg().c_str());
+    megdnn_assert(filter.ndim >= 4_z && filter.ndim <= 7_z, "%s", errmsg().c_str());
```
Comment (yeasoon) [inline dnn/src/common/convolution.cpp:1159]: =7_z 对应的情况是？

File: dnn/test/fallback/convolution.cpp [hunk 2]
```diff
@@ -707,4 +761,71 @@ TEST_F(FALLBACK, CONVOLUTION_BACKWARD_DATA_NAIVE_ALGO) {
...
+        size_t FW = fw;
+        size_t total_flops = IC * diff.total_nr_elems() * FH * FW * 2;
+        printf("nchw_time: %.3f ms  nchw_flops: %.3f Gflops\n", tnchw,
+               total_flops / (tnchw / RUN * 1e6));
+        printf("nchw44_time: %.3f ms  nchw44_flops: %.3f Gflops\n", tnchw44,
+               total_flops / (tnchw44 / RUN * 1e6));
+        printf("speedup: %.3f\n", tnchw / tnchw44);
+    };
+    run(1, 16, 14, 14, 16, 3, 3, 1, 1, 1, 1);
+    run(1, 32, 28, 28, 16, 3, 3, 1, 1, 1, 1);
```
Comment (yeasoon) [inline dnn/test/fallback/convolution.cpp:825]: 加几个hw 比较大以及多 batch的例子

File: dnn/src/common/convolution.cpp (hunk 1 above)
Comment (wangxiang9603) [reply dnn/src/common/convolution.cpp:1159]: group conv的情况，这时候filter为(g, oc/4, ic/4, fh, fw, 4, 4)

File: dnn/src/fallback/convolution/algos.cpp [hunk 3]
```diff
@@ -1,4 +1,5 @@
 #include "src/fallback/convolution/algos.h"
+#include "megdnn/opr_param_defs.h"
```
Comment (chenqy4933) [inline dnn/src/fallback/convolution/algos.cpp:2]: 这个头文件不用include

File: dnn/test/fallback/convolution.cpp [hunk 4]
```diff
@@ -707,4 +761,71 @@ TEST_F(FALLBACK, CONVOLUTION_BACKWARD_DATA_NAIVE_ALGO) {
...
+    run(1, 48, 28, 28, 48, 2, 2, 1, 0, 1, 1);
+    run(1, 32, 26, 26, 32, 3, 3, 1, 0, 1, 1);
```
[the remark this answers is not in the document]
Comment (wangxiang9603) [reply dnn/test/fallback/convolution.cpp:827]: nchw_time: 4.507 ms  nchw_flops: 10.020 Gflops
nchw44_time: 3.537 ms  nchw44_flops: 12.767 Gflops
speedup: 1.274
nchw_time: 29.032 ms  nchw_flops: 12.444 Gflops
nchw44_time: 17.445 ms  nchw44_flops: 20.709 Gflops
speedup: 1.664
nchw_time: 25.675 ms  nchw_flops: 28.142 Gflops
nchw44_time: 27.386 ms  nchw44_flops: 26.383 Gflops
speedup: 0.938
nchw_time: 23.208 ms  nchw_flops: 26.844 Gflops
nchw44_time: 24.900 ms  nchw44_flops: 25.020 Gflops
speedup: 0.932
nchw_time: 263.775 ms  nchw_flops: 28.622 Gflops
nchw44_time: 271.165 ms  nchw44_flops: 27.842 Gflops
speedup: 0.973
nchw_time: 262.505 ms  nchw_flops: 22.020 Gflops
nchw44_time: 256.677 ms  nchw44_flops: 22.520 Gflops
speedup: 1.023

[the remark this answers is not in the document]
Comment (yeasoon) [reply dnn/test/fallback/convolution.cpp:827]: 贴下测试的平台和测试指令？

**Comment by yeasoon** [comment]
others LGTM

**Comment by chenqy4933** [comment]
LGTM

**Pull request merged**